bdns/dns.go

```diff
@@ -488,37 +493,59 @@ func isPrivateV6(ip net.IP) bool {
 	return false
 }
 
-func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, error) {
-	resp, err := dnsClient.exchangeOne(ctx, hostname, ipType)
+func (dnsClient *impl) lookupIP(ctx context.Context, hostname string, ipType uint16) ([]dns.RR, string, error) {
+	resp, resolver, err := dnsClient.exchangeOne(ctx, hostname, ipType)
+	switch ipType {
+	case dns.TypeA:
+		if resolver != "" {
+			resolver = "A:" + resolver
+		}
+	case dns.TypeAAAA:
+		if resolver != "" {
+			resolver = "AAAA:" + resolver
+		}
+	}
 	errWrap := wrapErr(ipType, hostname, resp, err)
 	if errWrap != nil {
-		return nil, errWrap
+		return nil, resolver, errWrap
 	}
-	return resp.Answer, nil
+	return resp.Answer, resolver, nil
 }
 
 // LookupHost sends a DNS query to find all A and AAAA records associated with
 // the provided hostname. This method assumes that the external resolver will
 // chase CNAME/DNAME aliases and return relevant records. It will retry
 // requests in the case of temporary network errors. It returns an error if
 // both the A and AAAA lookups fail or are empty, but succeeds otherwise.
-func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+func (dnsClient *impl) LookupHost(ctx context.Context, hostname string) ([]net.IP, ResolverAddr, error) {
 	var recordsA, recordsAAAA []dns.RR
 	var errA, errAAAA error
+	var resolverA, resolverAAAA string
 	var wg sync.WaitGroup
 
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
+		recordsA, resolverA, errA = dnsClient.lookupIP(ctx, hostname, dns.TypeA)
 	}()
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		recordsAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
+		recordsAAAA, resolverAAAA, errAAAA = dnsClient.lookupIP(ctx, hostname, dns.TypeAAAA)
 	}()
 	wg.Wait()
 
+	// Each go func above could choose a different resolver to query for A and
+	// AAAA records. We filter out empty strings and otherwise join the
+	// resolvers into a ResolverAddr such as: A:127.0.0.1:53,AAAA:127.0.0.1:5353
+	var resolvers []string
+	resolvers = append(resolvers, resolverA, resolverAAAA)
+	resolvers = slices.DeleteFunc(resolvers, func(a string) bool {
+		return a == ""
+	})
+	slices.Sort(resolvers)
+	resolver := ResolverAddr(strings.Join(resolvers, ","))
```

Comment:
Comma+space could make this a bit easier to read in a log.

```suggestion
	// resolvers into a ResolverAddr such as: A:127.0.0.1:53, AAAA:127.0.0.1:5353
	var resolvers []string
	resolvers = append(resolvers, resolverA, resolverAAAA)
	resolvers = slices.DeleteFunc(resolvers, func(a string) bool {
		return a == ""
	})
	slices.Sort(resolvers)
	resolver := ResolverAddr(strings.Join(resolvers, ", "))
```

Reply:
I've addressed this an alternate way and have removed the `slices.DeleteFunc`.

Reply:
As a result of [this conversation](https://github.com/letsencrypt/boulder/pull/7271#discussion_r1475424351), I've gone back to using `slices.DeleteFunc`.